### PR TITLE
Add provision to register original declaration of interface functions as used

### LINKED ISSUE #1526: Divergence in output value returned from function call
```fortran
function exp() result(y)
real :: y(1)
y(1) = 1.0
end function exp

program main
real :: x(1)
interface 
function exp() result(y)
real :: y(1)
end function exp
end interface
x = exp()
print *, x
end program main
```

> GFortran

```console
$ gfortran a.f90 && ./a.out
   1.00000000    
```
> LFortran

```console
$ lfortran a.f90
0.00000000e+00 
```

### PR DESCRIPTION
**Pull request by Pranavchiku**
Fixes #1526.

### CODE REVIEW

File: integration_tests/functions_17.f90 [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+end function func
+
+program main
+real :: x(1)
+interface 
+function func() result(y)
+real :: y(1)
+end function func
+end interface
+x = func()
```
Comment (czgdp1807) [inline integration_tests/functions_17.f90:13]: What will happen if I remove this line? Will `func` be treated as unused interface?

Comment (Pranavchiku) [reply integration_tests/functions_17.f90:13]: Yes, as we will not have any function calls.

File: src/libasr/pass/unused_functions.cpp [hunk 2]
```diff
@@ -73,11 +73,33 @@ class CollectUnusedFunctionsVisitor :
...
                 fn_used[h] = name;
                 h = get_hash((ASR::asr_t*)x.m_name);
                 fn_used[h] = name;
+                if ( f_type->m_deftype == ASR::deftypeType::Interface ) {
+                    // find original declaration and add it to used
+                    SymbolTable *current_scope = f->m_symtab;
+                    // backtrace till we reach global scope or find the original declaration
+                    while (current_scope->parent) {
+                        current_scope = current_scope->parent;
+                        const ASR::symbol_t *s = current_scope->resolve_symbol(name);
```
Comment (certik) [inline src/libasr/pass/unused_functions.cpp:92]: `resolve_symbol` already goes recursively to parent scopes, so you can remove the while loop here, it is redundant.

**Comment by certik** [comment]
This logic seems fragile.

Rather, let's simply make sure that all functions in global scopes are not removed.

Later, we can link an interface with the actual function that will be used at ASR level, and then we can determine if a given global function is used or not.

File: tests/reference/julia-arrays_02-849d7b0.stdout [hunk 3]
```diff
@@ -1,893 +0,0 @@
...
-
-function reshape(x::Array{Int32, 1}, shape_vec::Array{Int32, 1})::Int32
-    local r::Int32 = 0
-    return r
-end
-
-function shape(x::Array{Int32, 1})::Int32
-    local shape::Int32 = 0
-    return shape
-end
```
Comment (Pranavchiku) [inline tests/reference/julia-arrays_02-849d7b0.stdout:84]: I am not sure why this `wat-*` references change, rest all are fine.

**Review by certik — APPROVED** [review]
I think that this is fine.

**Pull request merged**